src/EndpointV2/EndpointV2Middleware.php

```diff
@@ -303,4 +312,57 @@ private function normalizeAuthScheme(array $authScheme) : array
 
         return $normalizedAuthScheme;
     }
+
+    /**
+     * This method tries to resolve an AccountId parameter from a resolved identity.
+     * We will just perform this operation if the parameter AccountId is part of the ruleset parameters and
+     * AccountIdEndpointMode is not disabled, otherwise, we will ignore it.
+     * We also are going to resolve identity from a supplied credentials provider,
+     * and we will inject the resolved identity into the $command['@context'] property, so it can be reused along the way.
+     *
+     * @param array $commandArgs
+     * @param array $rulesetParams
+     * @return void
+     */
+    private function resolveAccountIdArgument(array &$commandArgs, array $rulesetParams)
+    {
+        static $ACCOUNT_ID_PARAM = 'AccountId';
+        $accountIdEndpointMode = $this->clientArgs['AccountIdEndpointMode'];
+        if (!isset($rulesetParams[$ACCOUNT_ID_PARAM]) || $accountIdEndpointMode === 'disabled') {
+            return;
+        }
+
+        $identity = null;
+        if (isset($commandArgs['@context']['resolved_identity'])) {
+            $identity = $commandArgs['@context']['resolved_identity']->wait();
+        } else {
+            $credentialsProviderFn = $this->credentialsProvider;
+            $identity = $credentialsProviderFn()->wait();
+            $commandArgs['@context']['resolved_identity'] = Create::promiseFor($identity);
```

Comment:
I think is safer to wrap the identity into a promise. I know we discussed that we are not logging $command in any of the middlewares, however, it could be done unintentionally in the future, or it could be appended by a custom customer implementation since the handler list is open for appending more middlewares. Wrapping the identity into a promise will make sure that if command is logged it will not log any credential values.

What do you think?

Reply:
Tested both and it's logged regardless of whether or not it's a promise.  The multi-auth PR contains some updates to the trace middleware which scrubs this from the logs.  I'd still like to keep `resolved_identity` "unwrapped" i.e. not contained within a promise